AdminServer/appscale/admin/base_handler.py

```diff
@@ -1,27 +1,165 @@
 """ A handler with helper functions that other handlers can extend. """
 
+import base64
+import hashlib
+import json
+import re
+import time
+
 from tornado import web
 from tornado.escape import json_encode
 from tornado.options import options
 
 from appscale.common.constants import HTTPCodes
+from appscale.common.ua_client import UAException
 from .constants import CustomHTTPError
 
 
 class BaseHandler(web.RequestHandler):
   """ A handler with helper functions that other handlers can extend. """
-  def authenticate(self):
+
+  # Scope for which users are authorized for AdminServer functionality.
+  AUTH_SCOPE = 'https://www.googleapis.com/auth/cloud-platform'
+
+  def authenticate(self, project_id, ua_client):
     """ Ensures requests are authenticated.
 
+    Args:
+      project_id: The project that we are authenticating for.
+      ua_client: A UA Client, used to check if the user can access the given
+        project.
+
     Raises:
-      CustomHTTPError if the secret is invalid.
+      CustomHTTPError if the secret or access token is invalid.
     """
-    if 'AppScale-Secret' not in self.request.headers:
-      message = 'A required header is missing: AppScale-Secret'
+    authorization_header = 'Authorization' in self.request.headers
+    secret_header = 'AppScale-Secret' in self.request.headers
+    if not authorization_header and not secret_header:
+      message = 'A required header is missing: AppScale-Secret or Authorization'
       raise CustomHTTPError(HTTPCodes.UNAUTHORIZED, message=message)
-
-    if self.request.headers['AppScale-Secret'] != options.secret:
+    if secret_header and self.request.headers['AppScale-Secret'] != \
+        options.secret:
       raise CustomHTTPError(HTTPCodes.UNAUTHORIZED, message='Invalid secret')
+    elif authorization_header:
+      try:
+        self.authenticate_access_token(
+          self.request.headers['Authorization'].split()[1],
+          project_id, ua_client)
+      except IndexError:
```

Comment:
I may have misled you on this one. I can't remember. From an API design perspective, it's a little unusual to have to handle an IndexError at this level. I would recommend raising the CustomHTTPError in the authenticate_access_token method.